crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -1451,37 +1457,53 @@ impl<'db> TypeInferenceBuilder<'db> {
             use_def.declarations_at_binding(binding)
         };
 
-        let declared_ty = symbol_from_declarations(self.db(), declarations)
-            .and_then(|symbol| {
-                let symbol = if matches!(symbol.symbol, Symbol::Type(_, Boundness::Bound)) {
-                    symbol
+        let declared_ty = place_from_declarations(self.db(), declarations)
+            .and_then(|place| {
+                Ok(if matches!(place.place, Place::Type(_, Boundness::Bound)) {
+                    place
                 } else if skip_non_global_scopes
                     || self.scope().file_scope_id(self.db()).is_global()
                 {
                     let module_type_declarations =
-                        module_type_implicit_global_declaration(self.db(), symbol_name)?;
-                    symbol.or_fall_back_to(self.db(), || module_type_declarations)
+                        module_type_implicit_global_declaration(self.db(), expr)?;
+                    place.or_fall_back_to(self.db(), || module_type_declarations)
                 } else {
-                    symbol
-                };
-                Ok(symbol)
-            })
-            .map(|SymbolAndQualifiers { symbol, .. }| {
-                symbol.ignore_possibly_unbound().unwrap_or(Type::unknown())
+                    place
+                })
             })
+            .map(
+                |PlaceAndQualifiers {
+                     place: resolved_place,
+                     ..
+                 }| {
+                    if resolved_place.is_unbound() && !place_table.place_expr(place_id).is_name() {
+                        if let Place::Type(ty, Boundness::Bound) = fallback_place(
+                            db,
+                            file_scope_id.to_scope_id(db, self.file()),
+                            place_table.place_expr(place_id),
+                        )
+                        .place
+                        {
+                            return ty;
+                        }
+                    }
+                    resolved_place
+                        .ignore_possibly_unbound()
+                        .unwrap_or(Type::unknown())
```

Comment:
Similar to my other comment: I would like to understand in what kind of situations this code kicks in. Can you give an example?